### ISSUE TITLE
Custom validator for unknown formats

### ISSUE THREAD

**Issue by dannydeut** [cdimascio/express-openapi-validator#210]
First of all, great module! It is working really well and saved us a lot of time.

There is an option `unknownFormats` to ignore custom formats. Is it possible to write your own validator and somehow inject it? Would be nice if I could use the base of this module instead of writing extra code to do same as this module does. Any other suggestions are welcome as well.

**Comment by cdimascio** [comment]
@dannydeut thank you. this is a great idea.
if you already have some ideas, i would love to hear your thoughts. particularly if you have ideas on how you envision the interface, etc.

**Comment by dannydeut** [comment]
@cdimascio Haven't really though about it. I guess you only need callback function that should returns success or an error message. As arguments I think you need the `value` and the matching object property with all its members.

**Comment by tkarls** [comment]
I too would like that.

I think instead of passing in an array with strings like ['formats', 'to', ignore'] one could pass in a map with format validators.

````
let validators = {
    myCustomFormat: value => {
        return true; // always valid
    },
    anotherFormat: value => {
        if (value === 'something') {
            throw Error('Value must be something else');
        }

        return true;
    }
};
```

The way I see it would be to get access to the value and then be able to either return either true or false. And by throwing an error the custom error message would be used.

I don't think it needs to be async. I think validating the format of a piece of data should be a straightforward thing. If you want to support async then returning a promise instead would be great.

**Comment by gn00616031** [comment]
ajv support this option 
**formats**: an object with custom formats. Keys and values will be passed to addFormat method.
So I suggest that you can add this option and pass to ajv, and it can do custom validation as well.

**Comment by ViniGodoy** [comment]
I upvote for this thread.

**Comment by rscarson** [comment]
A lot of custom formats could be validated automatically; if the pattern field was set, you could check against the enclosed regex to validate.

**Comment by rscarson** [comment]
Actually it seems ajv does that already for ignored formats! The ajvOptions parameter allows it to be passed through.

**Comment by cdimascio** [comment]
@rscarson yes. Good point. Would u be interested in putting up a PR?

https://github.com/ajv-validator/ajv#formats

The option can be added [here](https://github.com/cdimascio/express-openapi-validator/blob/master/src/openapi.validator.ts#L42)

Then pass the format option to the request and response validators [here](https://github.com/cdimascio/express-openapi-validator/blob/master/src/openapi.validator.ts#L271) and [here](https://github.com/cdimascio/express-openapi-validator/blob/master/src/openapi.validator.ts#L292)

The code is already setup to pass format from options to ajv. See [here](https://github.com/cdimascio/express-openapi-validator/blob/master/src/framework/ajv/index.ts#L31)

Basically we just have to expose a new format option and pass it to the req and res validators. And of course document it in the readme

**Comment by rscarson** [comment]
Sure! I'll write it up in the morning.
I'll use the ajv.addFormat function.

It can take regex strings or functions, so you'd give the express-openapi-validator options something like:

    options = {
         addFormats: {
             name: 'regex validator',
             othername: (val) => true
         }
    }

And it would pass them along.

If that sounds alright I'll submit the PR first thing tomorrow

Note; I could also look-ahead in the spec to find any unknown types, and add them but see [here](https://github.com/ajv-validator/ajv#redos-attack) for why that could be unsafe.

**Comment by cdimascio** [comment]
I've put up a [PR](https://github.com/cdimascio/express-openapi-validator/pull/436) for this. please have a look. interested in your feedback. thanks!

**Comment by cdimascio** [comment]
this feature is available in `v4.4.0`

e.g.

```javascript
formats: [
  {
    name: 'my-three-digit-format',
    type: 'number',
    // validate returns true the number has 3 digits, false otherwise
    validate: v => /^\d{3}$/.test(v.toString()) 
  },
  {
    name: 'my-three-letter-format',
    type: 'string',
    // validate returns true the string has 3 letters, false otherwise
    validate: v => /^[A-Za-z]{3}$/.test(v)
  },
]
```
